### PR TITLE
fix(memory): add OR fallback to search_like path

### LINKED ISSUE #2040: [TestAgent][agent-memory] bug: #1572 fix 不完整 — 长 prompt stride-sample 关键词含 2-char token 触发 search_like AND(无 OR fallback),auto-recall 注入 0 条
### Component

agent-memory

### Bug Description

#1572 的修复不完整:对长英文 stopwords prompt,adapter 的 `buildRecallQueries` 生成的 stride-sampled 关键词 candidate 首个 token 是 2-char `"um"`,触发 server 端 `search_like` 的 LIKE AND 路径(无 OR fallback),所有 4 个 candidate 都返回 0 结果,RRF 无可合并,auto-recall 注入 0 条。同时 #1405 的 OR fallback 修复只覆盖 BM25 path,未覆盖 `search_like` path,留下缝隙。

### Steps to Reproduce

任何能驱动 `agent-memory` 的 MCP 客户端都能复现(OpenClaw agent、stub plugin loader、或直接 JSON-RPC stdio)。在装有 agent-memory 0.2.5 + OpenClaw 的 ECS 上:

1. 在一个 namespace(`USER_ID=amtest-repro`)下用 `memory_observe` 落 3 条 seed 记忆:
   - M1(英文):`"Billing system runs on PostgreSQL 14. The slow query on the invoices table was optimized by adding a composite index on (customer_id, created_at). Codeword: CW1-REPRO."`
   - M2(中文):`"在 Kubernetes 集群中部署一个高可用的 PostgreSQL 数据库需要考虑 StatefulSet、PVC 持久化、Pod 反亲和性、以及 Secret 注入数据库密码。代号：CW2-REPRO。"`
   - M3(英文):`"The rate limiter uses a token bucket algorithm with a refill rate of 10 tokens/sec. Codeword: CW3-REPRO."`

2. 用下面这条 472-char 长 prompt 触发 `before_prompt_build` auto-recall hook(即 #1572 的 T2 场景):

```
so um i was wondering if you could maybe help me figure out something about the billing system you know the one that runs on postgresql and like we have this really slow query on the invoices table and it is causing all sorts of problems for the accounting team and i am pretty sure we need to optimize it but i am not really sure where to start with that and also there is this other thing about the database maybe needing some indexes or something like that anyway let me know what you think about all of this
```

3. 观察 auto-recall hook 返回的 `prependContext`:**空字符串**(`finalHits.length === 0`),M1 的 codeword `CW1-REPRO` 未被注入。

### Actual Behavior

`buildRecallQueries(T2)` 生成 4 个 candidate:

| # | candidate | len | 含 2-char token? | server 返回 |
|---|---|---|---|---|
| 0 | `um figure runs query problems pretty database let` | 49 | 是(`um`) | `[]` |
| 1 | `um figure runs` | 14 | 是(`um`) | `[]` |
| 2 | truncated 200 chars:`so um i was wondering if you could...` | 200 | 是(`um`/`so`/`i`/...) | `[]` |
| 3 | full 511 chars | 511 | 是 | `[]` |

所有 4 个 candidate 都返回空数组,RRF 无可合并条目,`finalHits.length === 0`,hook log:`agent-memory: auto-recall found 0 results for prompt (query len=511)`,`prependContext` 为空。

plugin log(无 warning,所有 candidate 都正常返回空数组,非 error/non-JSON):

```
[info] agent-memory: plugin registered (binary=/usr/bin/agent-memory, uid=amtest-repro, profile=advanced, session=ses_xxx)
[info] agent-memory: auto-recall found 0 results for prompt (query len=511)
```

### Expected Behavior

按 #1572 的修复意图(以及 #1572 修复对应的回归测试 T2 场景):长英文 stopwords prompt 应通过 `buildRecallQueries` 的多 candidate + RRF fusion 找到 M1(stride-sampled 关键词中 `runs`/`query` 的 trigram 应命中 M1 的 `runs`/`query`),把 M1 的 codeword 注入 `prependContext`。

### Root Cause

两个已 landed 的 fix 的交集缝隙:

**#1572 fix(adapter 端,`src/agent-memory/adapters/agent-memory/openclaw/src/keyword-extract.ts`)** 加了 `extractKeywords` + `buildRecallQueries`,对长 prompt stride-sample MAX_KEYWORDS=8 个 unique keyword。但 `extractKeywords` 保留 2-char token(注释说明 "While 2-char tokens cause the BM25 backend to fall back to LIKE search with AND semantics, they are important domain terms that should still be included")。对 T2(472 chars,32 个 unique keyword),stride-sample 的**首个** keyword 是 2-char `"um"`(unique keyword 列表的第 0 项),所有 4 个 candidate 都包含 `"um"`。

**#1405 fix(server 端,`src/agent-memory/src/index/store.rs` 的 `search_scoped`)** 给 BM25 path 加了 OR fallback(AND 返回 0 + tokens.len() > 1 时 retry OR-joined quoted tokens)。但 `search_scoped` 对**任一 token < 3 字符**的 query 走的是 `search_like`(LIKE AND)路径,**`search_like` 没有 OR fallback**:

```rust
// store.rs:336-345 (search_scoped)
let tokens: Vec<&str> = fts_q.split_whitespace().collect();
let needs_like_fallback = tokens.iter().any(|t| t.chars().count() < 3);
if needs_like_fallback {
    return self.search_like(&tokens, top_k, cold_filter, superseded_filter, &scope);
}

// store.rs:487-496 (search_like — 无 OR fallback)
let like_clause = tokens
    .iter()
    .map(|_| "files_fts.body LIKE ? ESCAPE '\\'")
    .collect::<Vec<_>>()
    .join(" AND ");  // ← AND,不是 OR;AND 返回 0 时无 retry
```

`"um"` 不在 M1/M2/M3 任何文件中,`body LIKE '%um%' AND body LIKE '%figure%' AND ...` 对所有文件返回 false,LIKE AND 查询 0 结果,且**不会 retry OR**。

所有 4 个 candidate(都含 `"um"`)都走 `search_like` AND 路径,都返回 0,RRF 无可合并,auto-recall 注入 0 条。

### Suggested Fix

任选其一即可让 T2 通过(推荐 A,更彻底):

**A. 给 `search_like` 加 OR fallback(对齐 BM25 path)**

```rust
// search_like 内,AND 返回 0 + tokens.len() > 1 时 retry OR-joined LIKE
let rows = if rows.is_empty() && tokens.len() > 1 {
    let or_clause = tokens
        .iter()
        .map(|_| "files_fts.body LIKE ? ESCAPE '\\'")
        .collect::<Vec<_>>()
        .join(" OR ");
    // ... retry with or_clause
} else {
    rows
};
```

**B. `extractKeywords` 在 unique keyword 数 > MAX_KEYWORDS 时只从 ≥3 字符的关键词 stride-sample**

2-char token 的"domain term"价值在短 prompt(几个 keyword 全用)时才体现;长 prompt stride-sample 时跳过 2-char token,避免触发 `search_like` AND。

```typescript
// keyword-extract.ts extractKeywords()
const longWords = unique.filter((w) => w.length >= 3);
if (longWords.length > MAX_KEYWORDS) {
  // stride-sample from longWords only, 2-char tokens dropped
  return Array.from({ length: MAX_KEYWORDS }, (_, i) =>
    longWords[Math.round((i * (longWords.length - 1)) / (MAX_KEYWORDS - 1))]
  ).join(" ");
}
return unique.join(" ");  // short prompt: keep 2-char tokens as before
```

### Environment

- OS: Anolis OS 8.x (kernel 6.6.102, x86_64)
- Component version: agent-memory 0.2.5
- Package/RPM: agent-memory-0.2.5-1.alnx4.x86_64
- OpenClaw: 2026.7.1-2

### Additional Context

- 相关 issue: #1572(CLOSED,#1572 fix 的 parent issue)、#1405(CLOSED,#1405 fix 给 BM25 path 加了 OR fallback 但未覆盖 `search_like`)
- 稳定性: 连续 3 次稳定复现(每次用独立 namespace)
- 触发条件: prompt 含 stopwords 比例高 + stride-sample 首个 unique keyword 是 2-char token(如 "um")。短 prompt(几个 keyword)不受影响,#1572 修复对短 prompt 有效。
- 这不是 #1572 的回归(#1572 fix 仍在),是 #1572 fix 对长 prompt + 2-char 首关键词场景的覆盖盲点 + #1405 fix 对 `search_like` path 的覆盖盲点。

### PR DESCRIPTION
**Pull request by ikunkun-sys**
## Description

Auto-recall keyword queries containing a 2-char token (e.g. `um`) route to the `search_like` LIKE AND path in `search_scoped`, which returned zero rows whenever any single token was absent from the corpus — the #1405 BM25 OR fallback never applied there, so long stopword-heavy prompts injected 0 memories (issue #2040, T2 scenario). This PR adds an OR-joined LIKE retry when the AND query finds nothing and the query has multiple tokens, mirroring the BM25 MATCH path. The server-side fix (option A from the issue) was chosen over dropping 2-char tokens in the adapter so all MCP clients benefit, not just OpenClaw.

## Related Issue

closes #2040

## Type of Change

- [x] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Refactoring (no functional change)
- [ ] Performance improvement
- [ ] CI/CD or build changes

## Scope

- [ ] `cosh` (copilot-shell)
- [ ] `cosh-ng` (cosh-ng)
- [ ] `sec-core` (agent-sec-core)
- [ ] `skill` (os-skills)
- [ ] `sight` (agentsight)
- [ ] `tokenless` (tokenless)
- [ ] `ckpt` (ws-ckpt)
- [x] `memory` (agent-memory)
- [ ] `anolisa` (anolisa-cli)
- [ ] `skillfs` (SkillFS)
- [ ] `blaze` (blaze)
- [ ] Multiple / Project-wide

## Checklist

- [x] I have read the [Contributing Guide](../CONTRIBUTING.md)
- [x] My code follows the project's code style
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have updated the documentation accordingly
- [ ] For `cosh`: Lint passes, type check passes, and tests pass
- [ ] For `cosh-ng`: `cargo clippy --all-targets -- -D warnings` and `cargo fmt --check` pass
- [ ] For `sec-core` (Rust): `cargo clippy -- -D warnings` and `cargo fmt --check` pass
- [ ] For `sec-core` (Python): Ruff format and pytest pass
- [ ] For `skill`: Skill directory structure is valid and shell scripts pass syntax check
- [ ] For `sight`: `cargo clippy -- -D warnings` and `cargo fmt --check` pass
- [ ] For `tokenless`: `cargo clippy -- -D warnings` and `cargo fmt --check` pass
- [x] For `memory` (Linux only): `cargo clippy --all-targets -- -D warnings`, `cargo fmt --check`, and `cargo test` pass
- [ ] For `anolisa`: `cargo clippy --all-targets --locked -- -D warnings`, `cargo fmt --all --check`, and `cargo test --locked` pass
- [ ] For `skillfs`: `cargo fmt --all --check`, `cargo clippy --workspace --all-targets -- -D warnings`, and `cargo test --workspace` pass
- [x] Lock files are up to date (`package-lock.json` / `Cargo.lock`)

## Testing

- Unit: new regression test `search_like_or_fallback_recalls_partial_matches` — multi-token LIKE-path query with a corpus-absent 2-char token (`um figure runs query`) recalls the partially matching document via OR fallback; single-token miss (`um`) stays empty (no OR retry possible).
- Full suite: `cargo fmt --all --check`, `cargo clippy --workspace --all-targets --locked -- -D warnings`, `cargo test --locked` — all pass (189 unit tests) on Linux x86_64.
- Verified the issue's claim independently: simulated the adapter's `extractKeywords` on the T2 prompt and reproduced the exact candidates (`um figure runs query problems pretty database let`, `um figure runs`) reported in #2040.
- Edge cases: existing `search_like_mixed_short_and_long_tokens` still passes (AND semantics preserved when AND finds rows); agent-scope, cold/superseded filter tests unaffected.

## Additional Notes

- The OR retry reuses the same parameterised bindings; the LIKE clause is parenthesised so OR terms don't bind looser than the trailing `is_cold` / `is_superseded` / agent-scope AND filters.
- Commit carries `Fixes: a89dbc3b` attribution to the #1405 fix that introduced the BM25-only OR fallback.

### CODE REVIEW

File: src/agent-memory/src/index/store.rs [hunk 1]
```diff
@@ -507,45 +510,62 @@ impl BM25Store {
...
-        );
-        // No SQL ORDER BY: the LIKE path has no rank, and ordering by mtime
-        // would discard high-frequency old documents before Rust scoring.
-        // Fetch a generous pool (top_k * 4) and let Rust sort by score +
-        // truncate, so the scoring dimension (frequency) — not mtime —
-        // decides what survives.
+            // Parenthesise the LIKE clause so OR-joined terms don't bind
+            // looser than the trailing AND filters.
+            let sql = format!(
+                r#"
```
[the remark this answers is not in the document]
Comment (ikunkun-sys) [reply src/agent-memory/src/index/store.rs]: Partially fixed in the latest revision (0ffe832b), with one factual correction to the finding:

**Correction**: the BM25 path's OR fallback does **not** use a 50-row pool — it uses `LIMIT top_k` directly (no over-fetch needed, since the SQL orders by `bm25()` rank). The fixed-50 pool only exists in `detect_conflicts_like`.

**Fixed (the valid core of the finding)**: the OR retry matches a superset of rows with no SQL-side ranking, so a `top_k*4` pool (20 by default) could fill with weak single-token matches and evict strong multi-token ones before Rust frequency scoring runs. The OR retry now uses `(top_k * 4).max(50)`, aligning with the `detect_conflicts_like` pool precedent.

**Not changed**: the AND branch keeps `top_k*4` — every row in an AND pool already matches all tokens, so the eviction risk the finding describes does not apply there.

[the remark this answers is not in the document]
Comment (kongche-jbw) [reply src/agent-memory/src/index/store.rs]: [P1] The 50-row floor still truncates candidates before relevance scoring.

I reproduced this against `0ffe832b`: insert 50 rows that match only the common token `is`, then insert `target.md` containing `needle` three times. A search for `is needle zz` with `top_k=5` returns only the weak rows. The absent `zz` makes the AND query empty, and the unordered OR query's `LIMIT 50` discards row 51 before the Rust frequency scoring can see it.

So `(top_k * 4).max(50)` lowers the probability but does not resolve the eviction problem. Please rank OR candidates before applying the limit (at least by matched-token count), or score all OR hits before truncating. A regression test with more than 50 weak matches would pin this behavior.

File: src/agent-memory/src/index/store.rs [hunk 2]
```diff
@@ -507,45 +510,66 @@ impl BM25Store {
...
+        // OR fallback mirroring the BM25 MATCH path: AND semantics return
+        // zero rows when any single token is absent from the corpus, which
+        // is routine for auto-recall keyword queries carrying filler tokens
+        // (e.g. "um"). Retry with OR so partial matches surface; the
+        // frequency scoring below still ranks multi-token hits first.
+        // OR matches a superset of rows with no SQL-side ranking, so a
+        // top_k*4 pool could fill with weak single-token matches before
+        // scoring runs — widen to at least 50 rows (the
+        // `detect_conflicts_like` pool precedent).
+        let rows = if rows.is_empty() && tokens.len() > 1 {
```
Comment (kongche-jbw) [inline src/agent-memory/src/index/store.rs:593]: [P1] Anchor OR-fallback snippets on a token present in each row

This branch can return a document that does not contain the globally shortest token (the reported query's `um` is exactly such a case), but the code below still passes that token to `make_snippet`. When it is absent, `make_snippet` returns the first 24 characters of the body.

I reproduced this with a document whose matching `needle` occurs after a long unrelated preamble and the query `um needle absentword`: the hit is returned, but its snippet is `"unrelated preamble longe"` and contains no matched content. The OpenClaw adapter injects only `hit.snippet`, so this does not actually recall the useful memory text. Please choose a needle per row from tokens that are actually present (with LIKE-compatible case handling), and extend the regression test to assert the snippet contains the matched term.

**Review by kongche-jbw — CHANGES_REQUESTED** [review]
Requesting changes for two correctness gaps in the current LIKE OR fallback:

- The fixed 50-row candidate pool can still evict the strongest partial match before Rust scoring; I added the concrete reproduction to the existing review thread.
- An OR result may not contain the globally shortest query token, so snippet generation can inject an unrelated body prefix instead of the matched memory content.

Both cases reproduce against `0ffe832b`. The existing regression test passes, but it only asserts the returned path and does not cover candidate eviction or the injected snippet.

**Review by Forrest-ly — COMMENTED** [review]
## Code Review Summary

**PR #2047** — fix(memory): add OR fallback to search_like path
**Commit**: `0ffe832`
**Reviewer**: Code Reviewer (AI)

### Overall Assessment

变更逻辑清晰，重构质量高。将 `run_query` 提取为闭包减少了 AND/OR 两条路径的代码重复，SQL 括号化处理正确避免了 OR 优先级问题。OR fallback 的条件守卫（`rows.is_empty() && tokens.len() > 1`）精确合理。回归测试覆盖了核心场景和边界情况。

总体而言，该 PR 是对 issue #2040 的有效修复，将 LIKE 路径从"完全无召回"提升到"部分召回"，与 BM25 MATCH 路径行为对齐。

### 发现的问题

以下两点与 @kongche-jbw 的 P1 意见一致，补充分析：

**[P1] OR fallback snippet 质量问题** (`store.rs:575-586`)

当 OR fallback 返回一个不包含全局最短 token（如 "um"）的文档时，`make_snippet` 的 `body.find(needle)` 返回 `None`，退化为截取 body 前 24 字符。这意味着用户虽然看到了匹配的文档路径，但 snippet 内容可能与查询无关。

建议修复：在 OR fallback 路径中，为每行选择该行实际包含的 token 作为 needle，而非使用全局最短 token。例如：
```rust
let needle = tokens.iter()
    .find(|t| body.to_lowercase().contains(&t.to_lowercase()))
    .copied()
    .unwrap_or("");
```

**[P1] OR 结果池截断风险** (`store.rs:569`)

`LIMIT (top_k * 4).max(50)` 配合无 `ORDER BY` 意味着超过 50 条匹配时截断顺序不确定。在极端场景下（大量弱匹配文档），最佳匹配可能被丢弃。

这比之前的零召回已有显著改善。如果要彻底解决，可考虑在 SQL 层增加简单的匹配 token 计数排序，或进一步扩大 pool。但考虑到 LIKE 查询无法利用 bm25() 排序，建议作为后续优化处理，不阻塞本次合并。

### 值得肯定的点

- 闭包重构消除了重复代码，使 AND/OR 路径共享同一执行逻辑
- SQL `WHERE ({like_clause})` 括号化正确处理了 OR 与后续 AND filter 的优先级
- `(top_k * 4).max(50)` 的池大小策略合理，引用了 `detect_conflicts_like` 的先例
- 回归测试同时覆盖了 OR fallback 命中和单 token 无匹配两个场景
- Commit message 的 `Fixes:` 和 `Closes` attribution 规范

### 结论

**COMMENT**（非阻塞）。建议优先修复 snippet 问题（改动小、影响大），结果池优化可作为后续改进。

File: src/agent-memory/src/index/store.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (ikunkun-sys) [reply src/agent-memory/src/index/store.rs]: Fixed in 4cac26ca. Confirmed the finding is correct — the 50-row floor only lowered the probability; an unordered LIMIT can still evict the strong row.

The OR retry now ranks in SQL before the LIMIT: `ORDER BY (body LIKE ?) + (body LIKE ?) + ... DESC` (LIKE evaluates to 0/1 in SQLite), so rows matching more tokens always enter the pool first; exact frequency scoring in Rust then decides the final order. The AND branch stays unordered — every AND row matches all tokens, so the count is a constant there.

Regression test added as suggested: `search_like_or_fallback_ranks_before_pool_truncation` — 60 weak rows (only `is` matches) + `target.md` inserted last; `is needle zz` with top_k=5 must return `target.md` first.

File: src/agent-memory/src/index/store.rs (hunk 2 above)
Comment (ikunkun-sys) [reply src/agent-memory/src/index/store.rs:593]: Fixed in 4cac26ca. Confirmed — OR-fallback rows may lack the globally shortest token, and `make_snippet` then degraded to the first 24 chars of the body, which is what the adapter injects.

The needle is now chosen per row: among query tokens actually present in that body (new `ascii_ci_find` helper — byte-wise ASCII case-insensitive to mirror SQLite LIKE semantics, exact bytes for non-ASCII/CJK), the shortest present one wins, and the body-cased occurrence is passed to `make_snippet` so the anchor works for mixed-case matches like `Needle`.

Regression test added as suggested: `search_like_or_fallback_snippet_anchors_on_present_token` — long unrelated preamble + `The Needle appears only here.`, query `um needle absentword`; asserts the snippet contains `«Needle»`.

**Review by kongche-jbw — APPROVED** [review]
lgtm

**Review by Forrest-ly — APPROVED** [review]
## Code Review: PR #2047 — fix(memory): add OR fallback to search_like path

**Commit**: `4cac26c`
**Reviewer**: Code Reviewer (agent)

### 总体评价：**APPROVE** ✅

This is a well-crafted, focused fix for the search recall gap identified in #2040. The changes are clean, well-documented, and the three regression tests cover the critical paths including the two findings from the previous review round.

### 审查详情

#### 正确性 ✓
1. **SQL 参数绑定顺序**：`run_query` 闭包内 WHERE LIKE patterns → agent param → ORDER BY LIKE patterns → LIMIT，与 SQL 文本 `WHERE ({like_clause}) {filters} {order_clause} LIMIT ?` 的位置参数顺序一致。正确。
2. **OR fallback 条件**：`rows.is_empty() && tokens.len() > 1` — 单 token 无需 OR 重试，正确。
3. **括号化**：`WHERE ({like_clause})` 防止 OR 项与后续 AND filters 优先级冲突。关键修复，正确。
4. **`ascii_ci_find` UTF-8 安全性**：ASCII 字节不会出现在 UTF-8 continuation byte 位置，byte-level `windows` 匹配不会跨越字符边界。注释准确。
5. **Snippet needle 选择**：从全局最短改为 per-row 最短存在 token，避免 OR fallback 行的 snippet 降级为 body 前 24 字符。

#### 安全性 ✓
所有查询使用参数化绑定（`?`），无 SQL 注入风险。Token 经 `sanitize_fts_query` 和 `like_pattern` 处理。

#### 性能 ✓
- OR fallback 仅在 AND 返回空时触发，额外开销有限
- `ascii_ci_find` 的 O(n*m) per row per token 受 pool size 约束
- ORDER BY matched-token-count 防止 LIMIT 截断强匹配行，有效解决了上轮 review 的 50-row floor 问题

#### 可维护性 ✓
- `run_query` 闭包消除代码重复，AND/OR 路径共享核心逻辑
- 注释充分解释 "why" 而非仅 "what"
- 命名清晰（`rank_matches`、`pool`、`joiner`）

#### 测试 ✓
三个回归测试覆盖核心场景：
1. `search_like_or_fallback_recalls_partial_matches` — OR fallback 基本行为 + 单 token 边界
2. `search_like_or_fallback_ranks_before_pool_truncation` — ORDER BY 保强匹配（60 弱行 + 1 强行的压力测试）
3. `search_like_or_fallback_snippet_anchors_on_present_token` — snippet 锚定在 body 中存在的 token

#### 文档 ✓
- 方法级 doc comment 更新，说明 OR fallback 语义
- 闭包内内联注释解释 ORDER BY 理由、参数绑定顺序、括号化必要性
- Commit message 包含 `Fixes:` attribution 和 `Closes #2040`

### 小建议（非阻塞）

1. **`run_query` 闭包重复计算 `like_patterns`**：闭包每次调用都重新 `tokens.iter().map(|t| like_pattern(t))` 生成 patterns。AND 和 OR 路径的 patterns 相同，可以提前计算并 capture。不过由于闭包最多调用两次，性能影响可忽略。

2. **`rows.flatten()` 静默跳过错误**：在闭包内 `rows.flatten().collect()` 会丢弃 SQLite 错误行。这是从原代码继承的行为，保持一致。如果未来需要更严格的错误处理，可以在这里加 `try_fold`。

### 结论

代码质量良好，修复了上轮 review（AGE-616 / commit `0ffe832`）提出的两个 P1 问题：
- [P1] 50-row floor 截断 → 已通过 ORDER BY + `.max(50)` pool 修复
- [P1] snippet 锚定缺失 → 已通过 per-row `ascii_ci_find` needle 选择修复

建议合并（已合并）。
